### ISSUE TITLE
[buildcff2vf] Cantarell VF corrupt

### ISSUE THREAD

**Issue by madig** [adobe-type-tools/afdko#657]
When blending Cantarell's CFF OTF masters into a CFF2 VF, the result does not pass ots-sanitize and several outlines are corrupt in ft-view (e.g. `D`).

1. `curl -O https://raw.githubusercontent.com/madig/cantarell-fonts/master/src/Cantarell.glyphs`
2. `fontmake -g Cantarell.glyphs -o otf --keep-overlaps`
3. `cp master_otf/*.otf master_ufo`
4. `buildcff2vf master_ufo/Cantarell.designspace`
5. `ots-sanitize master_ufo/Cantarell.otf`
```
ERROR at src/layout.cc:1387 (ParseDeviceTable)
ERROR: Layout: bad delta format: 32768
ERROR at src/gpos.cc:165 (ParseAnchorTable)
ERROR: GPOS: Failed to parse device table in anchor table
ERROR at src/gpos.cc:211 (ParseMarkArrayTable)
ERROR: GPOS: Faled to parse anchor table for mark table 0
ERROR at src/gpos.cc:648 (ParseMarkToAttachmentSubtables)
ERROR: GPOS: Failed to parse mark array
ERROR at src/layout.cc:1176 (Parse)
ERROR: Layout: Failed to parse lookup subtable 3
ERROR at src/layout.cc:238 (ParseLookupTable)
ERROR: Layout: Failed to parse subtable 0
ERROR at src/layout.cc:1326 (ParseLookupListTable)
ERROR: Layout: Failed to parse lookup 0
ERROR: GPOS: Failed to parse lookup list table
ERROR at src/ots.cc:675 (ProcessGeneric)
ERROR: GPOS: Failed to parse table
Failed to sanitize file!
```
Using `ft-view` to peer inside shows that several outlines are corrupt that displayed correctly before.

**Comment by anthrotype** [comment]
I belive this has to do with the CFF specializer in fontTools (which is on by default in the T2CharStringPen) that optimizes the CFF charstrings:

https://github.com/fonttools/fonttools/blob/62003f492fdd4b8c09ea15e526ba8540b8b3ab27/Lib/fontTools/pens/t2CharStringPen.py#L76-L82

@madig can you try to set `optimize=False` in this line?

https://github.com/googlei18n/ufo2ft/blob/20c766b80d2405df32e58c4488a3a6861b2e5ef7/Lib/ufo2ft/outlineCompiler.py#L898

you may also want to try with this https://github.com/fonttools/fonttools/pull/1345

**Comment by readroberts** [comment]
@madig The buildcff2vf from the current afdko requires the input CFF tables to be not optimized and not subroutinized. I suggest that after using fontmake to create the master_ufo directories, you use the afdko command 'buildmasterotfs master_ufo/Cantarell.designspace'  to build the master otf's, which creates the kind of OTFS that buildcff2vf expects.  At the moment, buildmasterotfs doesn't work because the UFO's are missing the 'postscriptFontName' from the fontinfo.plist. I was able to build this font without apparent problems after editing in the postscriptFontName values.

To use the new support for making CFF2 VF fonts in the proposed PR fonttools/fonttools#1345, you also need to use the afdko branch 'redesign_buildcff2'. However, I see that this also doesn't work with the Cantarell OTF's made by fontmake, I think because the otfs are subroutinized: you will still need to use the afdko 'buildmasterotfs'. This is a bug I will fix - part of the reason for the afdko 'redesign_buildcff2' branch is to work with any set of otfs that are topologically compatible.

**Comment by anthrotype** [comment]
fontmake has --no-subroutinize option. 
I'll change ufo2ft to also not run the CFF charstring specializer, whenever --no-subroutinize option is used.
I think the two go hand in hand. It's unlikely that one would like to do one without the other, so I prefer to reuse the same existing option than adding another one just for disabling the specialization.

**Comment by readroberts** [comment]
@madig The latest commit to  fonttools/fonttools#1345 fixes the bug I saw. With that commit in fonttools/fonttools#1345,  the buildcff2vf in the afdko branch 'redesign_buildcff2' can build your example font without having to turn off optimization and subroutinzation.

@anthrotype I think it is still useful to have the --no-subroutinize turn off both charstring optimization and subroutinzation. This does make the font much easier to inspect at a very low level.

**Comment by anthrotype** [comment]
> can build your example font without having to turn off optimization

the cffLib.specializer may merge or delete redundant operations, see:

https://github.com/fonttools/fonttools/blob/62003f492fdd4b8c09ea15e526ba8540b8b3ab27/Lib/fontTools/cffLib/specializer.py#L288-L291

so it's recommended that one turns it off when making CFF interpolatable.

**Comment by anthrotype** [comment]
it's probably `repr()` that adds the `u` in front of the `unicode` string in py

**Comment by anthrotype** [comment]
grr wwrong tab :blush:

**Comment by madig** [comment]
Just a quick note that `--optimize-cff 0 --keep-overlaps` in fontmake 1.8.0 still results in the same broken VF when building with `buildcff2vf` from the currently released afdko. Haven't tried the development branch yet.

**Comment by miguelsousa** [comment]
@madig do you mind testing this again. I was about to but https://raw.githubusercontent.com/madig/cantarell-fonts/master/src/Cantarell.glyphs gives me a 404. Thanks.

**Comment by madig** [comment]
I converted the sources to UFOs, the new sources are at https://github.com/madig/cantarell-fonts/tree/master/src.

A quick check with fontmake 2.0.1 (`-o otf-interpolatable`), psautohint 2.0.0a1 (`-r Cantarell-Regular.otf Cantarell-Bold.otf Cantarell-Light.otf`) and afdko 3.0.1 shows no obvious corruption in OTMaster 7.9 (though there are some things that might be source errors, haven't had time to check). The new CFF2 support for ots-sanitize still has to be released I think? Have to test with ft-view later.

**Comment by cjchapman** [comment]
@madig Khaled just released OTS 8.0.0 with CFF2 support a few hours ago:
https://github.com/khaledhosny/ots/releases/tag/v8.0.0

**Comment by madig** [comment]
(Thanks. Will test when I get back from holidays.)

**Comment by madig** [comment]
Ok, everything seems to be working fine. As in, if it's broken in the VF, it's broken in the statics as well :)

**Comment by readroberts** [comment]
@madig I cloned your branch of cantarell-fonts, and tried to follow the ReadMe instructions to. However, I got an error when I ran 'meson build': see below. Any advice?

 $ meson build
The Meson build system
Version: 0.52.0
Source dir: /Users/rroberts/adobe/github/opensource/cantarell-fonts
Build dir: /Users/rroberts/adobe/github/opensource/cantarell-fonts/build
Build type: native build
Project name: cantarell-fonts
Project version: 0.111
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/Users/rroberts/afdko-dev/bin/python3)
Program psautohint found: YES (/Users/rroberts/afdko-dev/bin/psautohint)
WARNING: Gettext not found, all translation targets will be ignored.

appstream/meson.build:3:0: ERROR: Tried to assign an invalid value to variable.

**Comment by readroberts** [comment]
When using the latest tools, the VF version of Cantarell VF passes ots-sanitizer, shows in fontviewcts, and snapshots with 'tx' to CFF instances that are identical to the master source fonts. No longer a bug.

I was able to build OTF versions of the Cantarell VF source fonts with 'python cantarell-fonts/scripts/make-static-fonts.py', after editing the script to set "remove_overlaps' to False. I then built the VF version with makecff2vf. I also hinted it with psautohint. Both the hinted and unhinted CFF2 VF font passed the tests above.
